content/blog/2025-12-16-loop-parallelization-proposal.md.md

```diff
@@ -0,0 +1,552 @@
+}
+```
+
+Fusion can reduce loop overhead and improve cache utilization. When combined with parallelization, it can also reduce the number of parallel region creations (fork/join overhead).
+
+## LLVM Infrastructure
+
+We built our pass using LLVM's modern pass infrastructure:
+- **LoopAccessAnalysis**: Performs memory dependence analysis
+- **ScalarEvolution**: Analyzes induction variables and trip counts
+- **OpenMPIRBuilder**: Generates OpenMP parallel constructs
+- **AliasAnalysis**: Determines memory aliasing relationships
+
+# Design and Implementation
+
+We implemented three variants to explore different optimization strategies:
+
+## Implementation 1: Original
+
+The original implementation focuses solely on parallelization. The core logic is the `isLoopParallelizable` function, which determines if a loop can be safely executed in parallel:
+
+```cpp
+bool isLoopParallelizable(Loop *L, ...) {
+    // 1. Check loop has proper structure (preheader and latch)
+    if (!L->getLoopPreheader() || !L->getLoopLatch())
+        return false;
+
+    // 2. Use LLVM's LoopAccessAnalysis for dependence checking
+    LoopAccessInfo LAI(L, &SE, &TTI, &TLI, &AA, &DT, &LI);
+    if (!LAI.canVectorizeMemory())
+        return false;
+
+    // 3. Verify zero memory dependencies between iterations
+    const MemoryDepChecker &DepChecker = LAI.getDepChecker();
+    const auto *Deps = DepChecker.getDependences();
+    if (Deps && !Deps->empty())
+        return false;
+
+    return true;
+}
+```
+
+### Understanding the Dependence Analysis
+
+**LoopAccessInfo (LAI)** is LLVM's built-in framework for analyzing loop dependencies. It answers the question: "Can different loop iterations access the same memory location in a way that creates data races?"
+
+LAI uses several LLVM analysis components:
+- **ScalarEvolution (SE)**: Analyzes how array indices evolve across iterations (e.g., recognizes `a[i]` has stride 1)
+- **AliasAnalysis (AA)**: Determines if two pointers might reference the same memory
+- **DominatorTree (DT) & LoopInfo (LI)**: Provide control-flow and loop structure information
+- **TargetTransformInfo (TTI) & TargetLibraryInfo (TLI)**: Hardware-specific and library function information
+
+**The key method** is `LAI.canVectorizeMemory()`, which returns true only when:
+1. All memory accesses have analyzable patterns
+2. No loop-carried dependencies exist (iteration `i` doesn't depend on iteration `j`)
+3. Memory accesses are safe to execute in any order
+
+**Example of safe loop:**
+```c
+for (int i = 0; i < n; i++)
+    c[i] = a[i] + b[i];  // Each iteration accesses different c[i]
+```
+
+**Example of unsafe loop (detected by LAI):**
+```c
+for (int i = 1; i < n; i++)
+    a[i] = a[i-1] + 1;   // Iteration i reads what iteration i-1 wrote
+```
+
+The `MemoryDepChecker` explicitly verifies no dependencies exist by checking that its internal dependency list is empty. This double-checking ensures we only parallelize loops that are provably safe.
```

Comment:
I don't think it's correct to say that this is something that `MemoryDepChecker` does; it's something that your code uses it to do, right?